bundle/src/main/java/com/adobe/acs/commons/wcm/properties/shared/impl/SharedComponentPropertiesReferenceProvider.java

```diff
@@ -0,0 +1,192 @@
+                componentResourceTypes.add(componentResourceType);
+            }
+
+            if (resource.hasChildren()) {
+                Iterator<Resource> children = resource.listChildren();
+                while (children.hasNext()) {
+                    componentResourceTypes.addAll(collectComponentResourceTypes(children.next()));
+                }
+            }
+        }
+        return componentResourceTypes;
+    }
+
+
+    /**
+     * Iterates through the given collection of resourceTypes and finds all the components that has
+     * shared dialog, then adds the root page of the given resource to the list.
+     *
+     * @param resource {@link Resource}
+     * @param componentResourceTypes {@link Set<String>}
+     * @return references
+     */
+    private List<Reference> findSharedComponents(Resource resource,
+        Set<String> componentResourceTypes) {
+        List<Reference> references = new ArrayList<>();
+
+        ResourceResolver resourceResolver = resource.getResourceResolver();
+        for (String resourceType : componentResourceTypes) {
+
+            Resource componentResource = getComponentResource(resourceType, resourceResolver);
+            if (componentResource == null) {
+                continue;
+            }
+
+            String jcrPrimaryType = componentResource.getValueMap().get(JcrConstants.JCR_PRIMARYTYPE, String.class);
+            if (StringUtils.isEmpty(jcrPrimaryType) || !jcrPrimaryType.equalsIgnoreCase("cq:Component")) {
+                continue;
+            }
+
+            Resource sharedDialog = componentResource.getResourceResolver()
+                .getResource(componentResource.getPath() + "/dialogshared");
+            Resource globalDialog = componentResource.getResourceResolver()
+                .getResource(componentResource.getPath() + "/dialogglobal");
+            if (sharedDialog == null && globalDialog == null) {
+                continue;
+            }
+
+            Page rootPage = pageRootProvider.getRootPage(resource);
+            if (rootPage != null) {
+                references.add(new Reference("sharedProperties", getReferenceName(rootPage),
+                    rootPage.getContentResource(),
+                    rootPage.getLastModified() != null ? rootPage.getLastModified()
+                        .getTimeInMillis() : -1L));
+
+                return references;
+            }
+        }
+        return references;
+    }
+
+    /**
+     * Returns the Component Resource by searching the configured resolver search paths ["/libs", "/apps"]
+     * @param resourceType {@link String}
+     * @param resourceResolver {@link ResourceResolver}
+     * @return Resource
+     */
+    private Resource getComponentResource(String resourceType, ResourceResolver resourceResolver) {
+        String[] resolverSearchPaths = resourceResolver.getSearchPath();
+
+        for (String searchPath : resolverSearchPaths) {
```

Comment:
Should we add logic here to see if `resourceType` starts with `/` and if so then fetch directly rather than by appending search path?  Though it is standard practice to use a relative path `sling:resourceType` such as `mysite/components/mything` the absolute path of `/apps/mysite/components/mything` generally works.